### ISSUE TITLE
neo3 ABI readonly methods

### ISSUE THREAD

**Issue by igormcoelho** [neo-project/neo#927]
Neo3 should provide readonly identifier on manifest. When these functions are demanded, two things happen: offchain services execute operations by themselves (unless forced to send to chain); and two, a readonly storage context is provided. This way, a readonly invoked method will.not be able to invoke a read write method.

**Comment by shargon** [comment]
Could you especify what methods, in order to implement this?

**Comment by igormcoelho** [comment]
The simplest example I have in mind is: `balanceOf`.

Suppose you do the appcall `contractX "balanceOf" myscripthash`. Neo3 would look into `contractX`manifest, and notice that `balanceOf` has `readonly` flag (similar to `safe` flag, or is it the same?). If it's readonly, the application engine is loaded in readonly mode, and syscall GetCurrentContext would fail, for example (it would require GetReadOnlyStorageContext). 

On C# devpack, it could be an attribute:
```cs
[ReadOnlyMethod]
BigInteger balanceOf(byte[] me)
```

**Comment by igormcoelho** [comment]
@erikzhang is it exactly the same concept as `safe` flag? if it's the same, I then propose to rename `safe` to `readonly`.

**Comment by erikzhang** [comment]
Same.

**Comment by shargon** [comment]
Maybe is better to call it `readonly`

**Comment by erikzhang** [comment]
Agree. And also we can move it to ABI.

**Comment by lock9** [comment]
@igormcoelho  isn't this the same as #829 ? We can keep the discussion here if you guys prefer

**Comment by igormcoelho** [comment]
No, we need to put this on NEP-3 Ammendment... another one.

**Comment by lock9** [comment]
@igormcoelho  could you describe the difference between this one and #829? Maybe we need to rename one of them because they look equal to me 😂

**Comment by igormcoelho** [comment]
Ok, perhaps this is why I didn't understand it at first :joy:

There are many situations where you could possibly "lock" state updates.

* The proposal here is to lock it during invocation of a **Neo Standard Method** (these methods with strings and parameters), which is already what Erik had proposed months ago (but by the name of `safe` methods). This is a lock specially related to ABI invocation of methods, and indirectly connected to NEP-3

* At #829 I saw a different perspective, for example, by locking methods when invoked **from inside a smart contract**. I mean, during a `CALL`, or an `APPCALL` (now `System.Call`). That one is a very different perspective, because it affects how neovm-appengine would process everything after jumps (or calls).

In short, if both are the same, good for us: we just need to rename `safe` to `readonly` and close two issues ;)

**Comment by igormcoelho** [comment]
No need to lock this feature to me. Focusing on other things first.

**Comment by lock9** [comment]
Sorry @igormcoelho . I assigned you because you said you would like to work on this task later. I suppose someone else can do this while you are working on other tasks?

**Comment by shargon** [comment]
@igormcoelho do you want to implement this?

**Comment by igormcoelho** [comment]
@shargon, I'm not sure that extending ExecutionContext is a good path... but I understand why you need this.
Perhaps we just need a flag field on ExecutionContext, and reuse it, for many things.

**Comment by shargon** [comment]
Today is oke flag, tomorrow two flags xD i don't like this model 😅

**Comment by igormcoelho** [comment]
Ok, it's approved on the vm side: https://github.com/neo-project/neo-vm/pull/191
We can advance on this one now, to validate those changes before merging.

**Comment by shargon** [comment]
Require https://github.com/neo-project/neo/pull/1048

**Comment by shargon** [comment]
This will be do o should be closed?

**Comment by shargon** [comment]
This will be do o should be closed? @neo-project/core

**Comment by lock9** [comment]
Hi @shargon, let's continue the discussion here: https://github.com/neo-project/proposals/pull/104
We need to update NEP-3 before continuing.
